url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/syntax/builder/SyntaxSeparation.java

```diff
@@ -0,0 +1,339 @@
+			if (tokenSymbols.get(o) != null)
+				throw new SyntaxException("Can not define ignored as token: "+o);
+		}
+		
+		boolean tokensWereDeclared = tokenSymbols.size() > 0;
+		List commandTokens = new ArrayList();
+		
+		// collect all `lexertokens`.
+		for (int i = 0; i < syntax.size(); i++)	{
+			Rule rule = syntax.getRule(i);
+
+			for (int j = 0; j < rule.rightSize(); j++)	{	// loop right side of rule
+				String sym = rule.getRightSymbol(j);
+				
+				if (sym.charAt(0) == Token.COMMAND_QUOTE)	{
+					sym = sym.substring(1, sym.length() - 1);	// remove command quotes
+					tokenSymbols.put(sym, sym);
+					commandTokens.add(sym);
+					rule.setRightSymbol(sym, j);	// remove the quotes for being able to recognize the symbol later
+				}
+				else
+				// get out all rules containing ".." and "-" when no "token" was defined explicitely, as these must be scanner rules
+				if (tokensWereDeclared == false && (sym.equals(Token.BUTNOT) || sym.equals(Token.UPTO)))	{
+					String s = rule.getNonterminal();	// take this rule to lexer rules
+					tokenSymbols.put(s, s);
+				}
+			}
+		}
+
+		// Get out all scanner rules from parser syntax, recursively
+		this.lexerSyntax = new Syntax(tokenSymbols.size() + ignoredSymbols.size());
+		Hashtable [] varr = new Hashtable [] { tokenSymbols, ignoredSymbols };
+		
+		for (int j = 0; j < varr.length; j++)	{	// loop token and ignored lists
+			Hashtable symbols = varr[j];
+			
+			for (Enumeration e = symbols.keys(); e.hasMoreElements(); )	{
+				String nonterm = (String)e.nextElement();
+				
+				getRulesUnderSymbol(nonterm, syntax, lexerSyntax, deleteIndexes);
+				
+				if (deleteIndexes.isEmpty() && lexerSyntax.hasRule(nonterm) == false)	{	// if there were no rules for symbol
+					String [][] predefinedRules = StandardLexerRules.rulesForIdentifier(nonterm);	// look for predefined rule
+					if (predefinedRules == null || predefinedRules.length <= 0)
+						throw new SyntaxException("Found nonterminal that has no rule and is no predefined lexer nonterminal: >"+nonterm+"<");
+					
+					lexerSyntax.appendRules(SyntaxUtil.ruleArrayToList(predefinedRules));
+				}
+				
+				deleteIndexes.removeIndexesFrom(syntax);
+			}
+		}
+		
+		// add ignoredSymbols to member variable lists
+		this.ignoredSymbols = new ArrayList(ignoredSymbols.size());
+		for (Enumeration e = ignoredSymbols.keys(); e.hasMoreElements(); )
+			this.ignoredSymbols.add(e.nextElement());
+
+		this.parserSyntax = provideParserSyntax(syntax, lexerSyntax, tokensWereDeclared, tokenSymbols, commandTokens, commandsDefinedAsTOKEN);
+
+		//System.err.println("Parser syntax:\n"+syntax);
+		//System.err.println("Lexer syntax:\n"+lexerSyntax);
+		//System.err.println("token symbols: "+tokenSymbols);
+		//System.err.println("ignored symbols: "+ignoredSymbols);
+	}
+
+
+	private Syntax provideParserSyntax(Syntax parserSyntax, Syntax lexerSyntax, boolean tokensWereDeclared, Map tokenSymbols, List commandTokens, List commandsDefinedAsTOKEN)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 24. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 250-262 into a separate method.